Content.Server/Speech/EntitySystems/IlleismAccentSystem.cs

```diff
@@ -0,0 +1,121 @@
+
+    // I/Myself -> NAME
+    private static readonly Regex RegexMyselfUpper = new(@"\bMYSELF\b");
+    private static readonly Regex RegexI = new(@"\bI\b|\bmyself\b", RegexOptions.IgnoreCase);
+
+    // Me -> NAME
+    private static readonly Regex RegexMeUpper = new(@"\bME\b");
+    private static readonly Regex RegexMeLower = new(@"\bme\b", RegexOptions.IgnoreCase);
+
+    // My crowbar -> NAME's crowbar
+    // That's mine! -> That's NAME's
+    private static readonly Regex RegexMyUpper = new(@"\bMY\b|\bMINE\b");
+    private static readonly Regex RegexMyLower = new(@"\bmy\b|\bmine\b", RegexOptions.IgnoreCase);
+
+    // I'll do it -> NAME'll do it
+    private static readonly Regex RegexIllUpper = new(@"\bI'?LL\b");
+    private static readonly Regex RegexIllLower = new(@"\bi'?ll\b", RegexOptions.IgnoreCase);
+
+
+    [Dependency] private readonly ReplacementAccentSystem _replacement = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<IlleismAccentComponent, AccentGetEvent>(OnAccent);
+    }
+
+    private bool MostlyUppercase(string message)
+    {
+        int totalLetters = 0;
+        int uppercaseLetters = 0;
+
+        // Iterate through each character in the string
+        foreach (char c in message)
+        {
+            if (char.IsLetter(c)) // Check if the character is a letter
+            {
+                totalLetters++;
+                if (char.IsUpper(c)) // Check if the letter is uppercase
+                {
+                    uppercaseLetters++;
+                }
+            }
+        }
+        if (totalLetters == 0)
+        {
+            return false;
+        }
+        return uppercaseLetters > totalLetters / 2;
+    }
+
+    private void OnAccent(EntityUid uid, IlleismAccentComponent component, AccentGetEvent args)
+    {
+        var entityManager = IoCManager.Resolve<IEntityManager>();
+        var message = args.Message;
+        var name = Name(uid).Split(' ')[0];
+
+        // I am going to Sec -> NAME is going to Sec
+        message = RegexIAmUpper.Replace(message, (name + " is").ToUpper());
+        message = RegexIAmLower.Replace(message, name + " is");
+
+        // I have it -> NAME has it
+        message = RegexIHaveUpper.Replace(message, (name + " has").ToUpper());
+        message = RegexIHaveLower.Replace(message, name + " has");
+
+        // I do! -> NAME does!
+        message = RegexIDoUpper.Replace(message, (name + " does").ToUpper());
+        message = RegexIDoLower.Replace(message, name + " does");
+
+        // I don't! -> NAME doesn't!
+        message = RegexIDontUpper.Replace(message, (name + " doesn't").ToUpper());
+        message = RegexIDontLower.Replace(message, name + " doesn't");
+
+        // I/myself -> NAME
+        message = RegexMyselfUpper.Replace(message, name.ToUpper());
+        if (MostlyUppercase(message))
+        {
+            message = RegexI.Replace(message, name.ToUpper());
```

Comment:
multiple calls ToUpper not only will result in method call but it also will allocate memory for string every time

Reply:
Added a var for the uppercase name and just made the strings uppercase rather than relying on ToUpper calls
Code now contains only one ToUpper call